[Partner] NVIDIA TRT Package

hinthornw opened the pull request:
Simplify #13976 and add as a separate package.

@jdye64 I've got some questions and need your help to validate that this works. Will add some comments below.

- [] Add README
- [X] Add doc notebook
- [X] Add simple LLM integration

Commit:
docs[patch]: fix databricks metadata (#14727)

lkuligin committed:
docs: updated branding for Google AI (#14728)

Replace this entire comment with:
  - **Description:** a small fix in branding

baskaryan committed:
infra: add integration test workflow (#14688)

baskaryan committed:
infra: Pre-release integration tests for partner pkgs (#14687)

baskaryan committed:
google-genai[patch]: add google-genai integration deps and extras (#14731)

Commit:
community[patch]: fix pgvector sqlalchemy (#14726)

Fixes #14699

baskaryan committed:
infra: add action checkout to pre-release-checks (#14732)

hinthornw committed:
Update vals

hinthornw commented on libs/partners/nvidia-trt/tests/integration_tests/test_llms.py:
```diff
@@ -0,0 +1,64 @@
+from langchain_nvidia_trt.llms import TritonTensorRTLLM
+
+
+def test_stream() -> None:
```
I'll have to add skip marks here since we can't run..

hinthornw commented on libs/partners/nvidia-trt/langchain_nvidia_trt/llms.py:
```diff
@@ -0,0 +1,405 @@
+        self._load_model(self.model_name)
+        # stop_words = stop if stop is not None else self.stop
+        generations = []
+        # TODO: We should handle the native batching instead.
```
@jdye64, I'm mainly following your structure here: I'm assuming TRT natively handles batching. Got any pointers?

hinthornw commented on libs/partners/nvidia-trt/langchain_nvidia_trt/llms.py:
```diff
@@ -0,0 +1,405 @@
+        # TODO: We should handle the native batching instead.
+        for prompt in prompts:
+            invoc_params = {**invocation_params, "prompt": [[prompt]]}
+            # request_id = str(random.randint(1, 9999999))  # nosec
```
@jdye64 it doesn't look like `client.infer` requires/accepts a request_id. Could you confirm that's only used for streaming?

jdye64 replied:
Correct. The `request_id` is used to manage GRPC streams on the Triton server end so certain API calls require it. I have refactored and removed the call to `client.infer` however since we are always using the GRPC client and not HTTP client now it is safer to just always use a stream and batch locally if needed. This is required because certain models and configurations on the Triton GRCP server side might cause calls to just `infer` to fail and instead need `async_stream_infer` calls.

hinthornw replied:
helpful - thanks!

hinthornw commented on libs/partners/nvidia-trt/langchain_nvidia_trt/llms.py:
```diff
@@ -0,0 +1,405 @@
+        invocation_params = self._get_invocation_params(**kwargs)
+        # (WFH) TODO: It looks like we were doing this already but...?
+        self._load_model(self.model_name)
+        # stop_words = stop if stop is not None else self.stop
```
For non-streaming invocations, I'm assuming we'll just strip off the suffix manually based on the stop words, but could you confirm this?

hinthornw commented on libs/partners/nvidia-trt/langchain_nvidia_trt/llms.py:
```diff
@@ -0,0 +1,405 @@
+        """Trim batch response by removing prompt and extra generated text."""
+        # extract the generated part of the prompt
+        # TODO: This assumes llama-style prompting...
+        split = result_str.split("[/INST]", 1)
```
@jdye64 , this is copied over from the other PR. It assumes llama-style prompting, which seems like a bad assumption. Could you confirm it's OK for me instead to remove the input prefix? Or will that be mangled in the response / unreliable?

I'm thinking basiclaly
```
self. _trim_batch_response(result_str, prompt_str)
...
```

hinthornw replied:
@jdye64 a followup:  if we're expecting this type of chat templating, should(n't) we be creating a chat model rather than an LLM interface? That way we can handle the structure on behalf of the user, rather than expecting them to do their own string formatting. That would also make it a lot more interoperable with the rest of langchain

hinthornw committed:
Rm httpclient import

hinthornw committed:
Add mypy ini

hinthornw commented on libs/partners/nvidia-trt/langchain_nvidia_trt/llms.py:
```diff
@@ -0,0 +1,404 @@
+                # the very last response might have no output, just the final flag
+                response = self._process_result(response_raw)
+                if force_batch:
+                    response = self._trim_batch_response(response)
```
For the streaming case, is it returning only new tokens? Or is it returning the full sequence on each iteration? If the former, this trim method will break (it looks for the input prompt/split token)

hinthornw committed:
Revert "[Hub|tracing] Tag hub prompts" (#14735)

Reverts langchain-ai/langchain#14720

baskaryan committed:
core[patch]: Release 0.1.1 (#14738)

Commit:
infra: docs build install community editable (#14739)

baskaryan committed:
infra: fix pre-release integration test and add unit test (#14742)

grisaitis committed:
docs: Remove trailing "`" in pip install command (#14730)

hi! just a simple typo fix in the local LLM python docs

- **Description:** removing a trailing "\`" character in a `!pip install
...` command
  - **Issue:** n/a
  - **Dependencies:** n/a
  - **Tag maintainer:** n/a
  - **Twitter handle:** n/a

lkuligin committed:
google-genai[patch], community[patch]: Added support for new Google GenerativeAI models (#14530)

Replace this entire comment with:
  - **Description:** added support for new Google GenerativeAI models
  - **Twitter handle:** lkuligin

---------

Co-authored-by: Erick Friis <erick@langchain.dev>

hinthornw committed:
[Tracing] String Stacktrace (#14131)

Add full stacktrace

hinthornw committed:
[Evals] End project (#14324)

Also does some cleanup.

Now that we support updating/ending projects, do this automatically.
Then you can edit the name of the project in the app.

hinthornw committed:
Fix OAI Tool Message (#14746)

See format here:
https://platform.openai.com/docs/guides/function-calling/parallel-function-calling


It expects a "name" argument, which we aren't providing by default.


![image](https://github.com/langchain-ai/langchain/assets/13333726/7cd82978-337c-40a1-b099-3bb25cd57eb4)


Alternative is to add the 'name' field directly to the message if people
prefer.

hinthornw committed:
Update propositional-retrieval template (#14766)

More descriptive name. Add parser in ingest. Update image link

hinthornw committed:
[Documentation] Updates to NVIDIA Playground/Foundation Model naming.… (#14770)

…  (#14723)

- **Description:** Minor updates per marketing requests. Namely, name
decisions (AI Foundation Models / AI Playground)
  - **Tag maintainer:** @hinthornw 

Do want to pass around the PR for a bit and ask a few more marketing
questions before merge, but just want to make sure I'm not working in a
vacuum. No major changes to code functionality intended; the PR should
be for documentation and only minor tweaks.

Note: QA model is a bit borked across staging/prod right now. Relevant
teams have been informed and are looking into it, and I'm placeholdered
the response to that of a working version in the notebook.

Co-authored-by: Vadim Kudlay <32310964+VKudlay@users.noreply.github.com>

lalanikarim committed:
community[minor]: Add SurrealDB vectorstore (#13331)

**Description:** Vectorstore implementation around
[SurrealDB](https://www.surrealdb.com)

---------

Co-authored-by: Bagatur <baskaryan@gmail.com>

hwchase17 committed:
langchain[patch]: remove unused imports (#14680)

Co-authored-by: Bagatur <baskaryan@gmail.com>

leo-gan committed:
docs: `Steam` update (#14778)

Updated the page title. It was inconsistent.
Updated page with links; description and setting details.

leo-gan committed:
docs: `cloudflare` update (#14779)

Added provider page.
Added links, descriptions

rlancemartin committed:
Add image support for Ollama (#14713)

Support [LLaVA](https://ollama.ai/library/llava):
* Upgrade Ollama
* `ollama pull llava`

Ensure compatibility with [image prompt
template](https://github.com/langchain-ai/langchain/pull/14263)

---------

Co-authored-by: jacoblee93 <jacoblee93@gmail.com>

leo-gan committed:
docs: `google drive` update (#14781)

The [Google Drive
toolkit](https://python.langchain.com/docs/integrations/toolkits/google_drive)
page is a duplicate of the [Google Drive
tool](https://python.langchain.com/docs/integrations/tools/google_drive)
page.
- Removed the `Google Drive toolkit` page (it shouldn't be a toolkit but
tool)
- Removed the correspondent reference in the Google platform page
- Redirected the removed page to the tool page.

tyumentsev4 committed:
community[patch]: Update YandexGPT API (#14773)

Update LLMand Chat model to use new api version

---------

Co-authored-by: Dmitry Tyumentsev <dmitry.tyumentsev@raftds.com>

NoahStapp committed:
community[patch]: Implement similarity_score_threshold for MongoDB Vector Store (#14740)

Adds the option for `similarity_score_threshold` when using
`MongoDBAtlasVectorSearch` as a vector store retriever.

Example use:

```
vector_search = MongoDBAtlasVectorSearch.from_documents(...)

qa_retriever = vector_search.as_retriever(
    search_type="similarity_score_threshold",
    search_kwargs={
        "score_threshold": 0.5,
    }
)

qa = RetrievalQA.from_chain_type(
	llm=OpenAI(), 
	chain_type="stuff", 
	retriever=qa_retriever,
)

docs = qa({"query": "..."})
```

I've tested this feature locally, using a MongoDB Atlas Cluster with a
vector search index.

Commit:
docs[patch]: fix zoom (#14786)

not sure why quarto is removing divs

hinthornw committed:
Permit updates in indexing (#14482)

Commit:
docs: developer docs (#14776)

Builds out a developer documentation section in the docs

- Links it from contributing.md
- Adds an initial guide on how to contribute an integration

---------

Co-authored-by: Bagatur <baskaryan@gmail.com>

Commit:
infra: cut down on integration steps (#14785)

<!-- Thank you for contributing to LangChain!

Replace this entire comment with:
  - **Description:** a description of the change, 
  - **Issue:** the issue # it fixes (if applicable),
  - **Dependencies:** any dependencies required for this change,
- **Tag maintainer:** for a quicker response, tag the relevant
maintainer (see below),
- **Twitter handle:** we announce bigger features on Twitter. If your PR
gets announced, and you'd like a mention, we'll gladly shout you out!

Please make sure your PR is passing linting and testing before
submitting. Run `make format`, `make lint` and `make test` to check this
locally.

See contribution guidelines for more information on how to write/run
tests, lint, etc:

https://github.com/langchain-ai/langchain/blob/master/.github/CONTRIBUTING.md

If you're adding a new integration, please include:
1. a test for the integration, preferably unit tests that do not rely on
network access,
2. an example notebook showing its use. It lives in `docs/extras`
directory.

If no one reviews your PR within a few days, please @-mention one of
@baskaryan, @eyurtsev, @hwchase17.
 -->

---------

Co-authored-by: Bagatur <baskaryan@gmail.com>

developer20sujeet committed:
community[patch]: support for Sybase SQL anywhere added. (#14821)

- **Description:** support for Sybase SQL anywhere added in
sql_database.py file at path
langchain\libs\community\langchain_community\utilities
- **Issue:** It will resolve default schema setting for Sybase SQL
anywhere
  - **Dependencies:** No,
  - **Tag maintainer:** @baskaryan, @eyurtsev, @hwchase17,
  - **Twitter handle:** NA

---------

Co-authored-by: learn360sujeet <121271779+learn360sujeet@users.noreply.github.com>
Co-authored-by: Bagatur <baskaryan@gmail.com>

tyumentsev4 committed:
community[patch]: fix agenerate return value (#14815)

Fixed:
  -  `_agenerate` return value in the YandexGPT Chat Model
  - duplicate line in the documentation

Co-authored-by: Dmitry Tyumentsev <dmitry.tyumentsev@raftds.com>

yacine555 committed:
docs: ensure consistency in declaring LANGCHAIN_API_KEY... (#14823)

... variable, accompanied by a quote

Co-authored-by: Yacine Bouakkaz <Yacine.Bouakkaz@evokegroup.com>

leo-gan committed:
docs redundant pages (#14774)

[ScaNN](https://python.langchain.com/docs/integrations/providers/scann)
and
[DynamoDB](https://python.langchain.com/docs/integrations/platforms/aws#aws-dynamodb)
pages in `providers` are redundant because we have those references in
the Google and AWS platform pages. It is confusing.
- I removed unnecessary pages, redirected files to new nams;

xkra committed:
docs: Typo in Templates README.md (#14812)

Corrected path reference from package/pirate-speak to
packages/pirate-speak

hinthornw committed:
community: Add logprobs in gen output (#14826)

Now that it's supported again for OAI chat models .

Shame this wouldn't include it in the `.invoke()` output though (it's
not included in the message itself). Would need to do a follow-up for
that to be the case

eriknw committed:
docs: Fix link typo to `/docs/integrations/text_embedding/nvidia_ai_endpoints` (#14827)

This page doesn't exist:
-
https://python.langchain.com/docs/integrations/text_embeddings/nvidia_ai_endpoints

but this one does:
-
https://python.langchain.com/docs/integrations/text_embedding/nvidia_ai_endpoints

Falimonda committed:
docs: typo in rag use case (#14800)

Description: Fixes minor typo to documentation

peterj committed:
docs: Fix the broken link to Extraction page (#14806)

**Description:** fixing a broken link to the extraction doc page

vladkol committed:
community[minor]: New model parameters and dynamic batching for VertexAIEmbeddings (#13999)

- **Description:** VertexAIEmbeddings performance improvements
  - **Twitter handle:** @vladkol

## Improvements

- Dynamic batch size, starting from 250, lowering down to 5. Batch size
varies across regions.
Some regions support larger batches, and it significantly improves
performance.
When running large batches of texts in `us-central1`, performance gain
can be up to 3.5x.
The dynamic batching also makes sure every batch is below 20K token
limit.
- New model parameter `embeddings_type` that translates to `task_type`
parameter of the API. Newer model versions support [different embeddings
task
types](https://cloud.google.com/vertex-ai/docs/generative-ai/embeddings/get-text-embeddings#api_changes_to_models_released_on_or_after_august_2023).

hinthornw committed:
Update parser (#14831)

Gpt-3.5 sometimes calls with empty string arguments instead of `{}`

I'd assume it's because the typescript representation on their backend
makes it a bit ambiguous.

hinthornw committed:
[Bugfix] Ensure tool output is a str, for OAI Assistant (#14830)

Tool outputs have to be strings apparently. Ensure they are formatted
correctly before passing as intermediate steps.
 

```
BadRequestError: Error code: 400 - {'error': {'message': '1 validation error for Request\nbody -> tool_outputs -> 0 -> output\n  str type expected (type=type_error.str)', 'type': 'invalid_request_error', 'param': None, 'code': None}}
```

169 committed:
community[patch]: Update Tongyi default model_name (#14844)

<img width="1305" alt="Screenshot 2023-12-18 at 9 54 01 PM"
src="https://github.com/langchain-ai/langchain/assets/10000925/c943fd81-cd48-46eb-8dff-4680424d9ba9">

The current model is no longer available.

VKudlay committed:
docs: update NVIDIA integration (#14780)

- **Description:** Modification of descriptions for marketing purposes
and transitioning towards `platforms` directory if possible.
- **Issue:** Some marketing opportunities, lodging PR and awaiting later
discussions.
  - 

This PR is intended to be merged when decisions settle/hopefully after
further considerations. Submitting as Draft for now. Nobody @'d yet.

---------

Co-authored-by: Bagatur <baskaryan@gmail.com>

jdye64 committed:
[Partner] Nvidia Triton updates for PR (#14761)

@hinthornw PR for branch `wfh/partners/trt` and PR
https://github.com/langchain-ai/langchain/pull/14733

Commit:
docs[patch]: gemini keywords (#14856)

Commit:
docs[patch]: more keywords (#14858)

baskaryan committed:
community[patch]: Release 0.0.4 (#14864)

baskaryan committed:
langchain[patch]: Release 0.0.351 (#14867)

hwchase17 committed:
add methods to deserialize prompts that were old (#14857)

Commit:
community: replace deprecated davinci models (#14860)

This is technically a breaking change because it'll switch out default
models from `text-davinci-003` to `gpt-3.5-turbo-instruct`, but OpenAI
is shutting off those endpoints on 1/4 anyways.

Feels less disruptive to switch out the default instead.

hwchase17 committed:
WIP: sql research assistant (#14240)

leo-gan committed:
docstrings `core` update (#14871)

Added missed docstrings

sweetkane committed:
Fix token text splitter duplicates (#14848)

- **Description:** 
- Add a break case to `text_splitter.py::split_text_on_tokens()` to
avoid unwanted item at the end of result.
    - Add a testcase to enforce the behavior.
  - **Issue:** 
    - #14649 
    - #5897
  - **Dependencies:** n/a,
 
---

**Quick illustration of change:**

```
text = "foo bar baz 123"

tokenizer = Tokenizer(
        chunk_overlap=3,
        tokens_per_chunk=7
)

output = split_text_on_tokens(text=text, tokenizer=tokenizer)
```
output before change: `["foo bar", "bar baz", "baz 123", "123"]`
output after change: `["foo bar", "bar baz", "baz 123"]`

leo-gan committed:
docstrings `langchain` update (#14870)

Added missed docstrings

soulhat committed:
docs: fixed tiktoken link error (#14840)

<!-- Thank you for contributing to LangChain!

Replace this entire comment with:
  - **Description:** fixed tiktoken link error, 
  - **Issue:** no,
  - **Dependencies:** no,
- **Tag maintainer:** for a quicker response, tag the relevant
maintainer (see below),
  - **Twitter handle:** no!

Please make sure your PR is passing linting and testing before
submitting. Run `make format`, `make lint` and `make test` to check this
locally.

See contribution guidelines for more information on how to write/run
tests, lint, etc:
https://python.langchain.com/docs/contributing/

If you're adding a new integration, please include:
1. a test for the integration, preferably unit tests that do not rely on
network access,
2. an example notebook showing its use. It lives in `docs/extras`
directory.

If no one reviews your PR within a few days, please @-mention one of
@baskaryan, @eyurtsev, @hwchase17.
 -->
- **Description:** fixed tiktoken link error, 
- **Issue:** no,
- **Dependencies:** no,
- **Tag maintainer:** @baskaryan,
- **Twitter handle:** SignetCode!

malandis committed:
docs: fix typo in contributing re installing integration test deps (#14861)

**Description**

The contributing docs lists a poetry command to install community for
dev work that includes a poetry group called `integration_tests`. This
is a mistake: the poetry group for integration tests is called
`test_integration`, not `integration_tests`. See here:

https://github.com/langchain-ai/langchain/blob/master/libs/community/pyproject.toml#L119

abhjaw committed:
Update kendra.py to avoid Kendra query ValidationException (#14866)

Fixing issue - https://github.com/langchain-ai/langchain/issues/14494 to
avoid Kendra query ValidationException

<!-- Thank you for contributing to LangChain!

Replace this entire comment with:
- **Description:** Update kendra.py to avoid Kendra query
ValidationException,
- **Issue:** the issue
#https://github.com/langchain-ai/langchain/issues/14494,
  - **Dependencies:** None,
  - **Tag maintainer:** ,
  - **Twitter handle:** 

If no one reviews your PR within a few days, please @-mention one of
@baskaryan, @eyurtsev, @hwchase17.
 -->

---------

Co-authored-by: Harrison Chase <hw.chase.17@gmail.com>

hwchase17 committed:
Harrison/agent docs custom (#14877)

asofter committed:
Improve prompt injection detection (#14842)

- **Description:** This is addition to [my previous
PR](https://github.com/langchain-ai/langchain/pull/13930) with
improvements to flexibility allowing different models and notebook to
use ONNX runtime for faster speed. Since the last PR, [our
model](https://huggingface.co/laiyer/deberta-v3-base-prompt-injection)
got more than 660k downloads, and with the [public
benchmark](https://huggingface.co/spaces/laiyer/prompt-injection-benchmark)
showed much fewer false-positives than the previous one from deepset.
Additionally, on the ONNX runtime, it can be running 3x faster on the
CPU, which might be handy for builders using Langchain.
 **Issue:** N/A
 - **Dependencies:** N/A
 - **Tag maintainer:** N/A 
- **Twitter handle:** `@laiyer_ai`

sevendark committed:
OPENAI_PROXY not working (#14833)

Replace this entire comment with:
- **Description:** OPENAI_PROXY is not working for openai==1.3.9, The
`proxies` argument is deprecated. The `http_client` argument should be
passed instead,
  - **Issue:** OPENAI_PROXY is not working,
  - **Dependencies:** None,
  - **Tag maintainer:** @hwchase17 ,
  - **Twitter handle:** timothy66666

leo-gan committed:
Docs `tencent` pages update (#14879)

- updated `Tencent` provider page: added a chat model and document
loader references; company description
- updated Chat model and Document loader pages with descriptions, links
- renamed files to consistent formats; redirected file names
Note:
I was getting this linting error on code that **was not changed in my
PR**!

> Error:
docs/docs/guides/safety/hugging_face_prompt_injection.ipynb:1:1: I001
Import block is un-sorted or un-formatted
> make: *** [Makefile:47: lint_package] Error 1

I've fixed this error in the notebook

lkuligin committed:
added history and support for system_message as param (#14824)

- **Description:** added support for chat_history for Google
GenerativeAI (to actually use the `chat` API) plus since Gemini
currently doesn't have a support for SystemMessage, added support for it
only if a user provides additional `convert_system_message_to_human`
flag during model initialization (in this case, SystemMessage would be
prepanded to the first HumanMessage)
  - **Issue:** #14710 
  - **Dependencies:** any dependencies required for this change,
- **Tag maintainer:** for a quicker response, tag the relevant
maintainer (see below),
  - **Twitter handle:** lkuligin

---------

Co-authored-by: William FH <13333726+hinthornw@users.noreply.github.com>

hinthornw committed:
[Partner] Google GenAi new release (#14882)

to support the system message merging

Also fix integration tests that weren't passing

jdye64 committed:
Nvidia Triton TRT updates for base branch wfh/partnets/trt (#14862)

@hinthornw more changes for
https://github.com/langchain-ai/langchain/pull/14733

hinthornw committed:
[Partner] Update google integration test (#14883)

Gemini has decided that pickle rick is unsafe:
https://github.com/langchain-ai/langchain/actions/runs/7256642294/job/19769249444#step:8:189


![image](https://github.com/langchain-ai/langchain/assets/13333726/cfbf4312-53b6-4290-84ee-6ce0742e739e)

hinthornw committed:
Merge branch 'master' into wfh/partners/trt

hinthornw committed:
skip int tests
